web/source/kmwembedded.ts

```diff
@@ -375,7 +375,7 @@ namespace com.keyman.text {
       
       // Note:  this assumes Lelem is properly attached and has an element interface.
       // Currently true in the Android and iOS apps.
-      var Lelem=keymanweb.domManager.getLastActiveElement(),keyShiftState=core.keyboardProcessor.getModifierState(layer);
+      var Lelem=keymanweb.domManager.getLastActiveElement(),keyShiftState=com.keyman.text.KeyboardProcessor.getModifierState(layer);
```

Comment:
I'm no longer clear on which globals we have. Is `com` a global?

Reply:
Namespaced code's root namespace is always a global, yes.

Reply:
When was `com` introduced? That was a while ago, am I right? Same issues exist with `window.com` as with `window.module` ... which I wasn't concerned about with `window.KeymanWeb`...

Is `com` documented for use by other devs or is internal to KMW? If it is internal, can we work to eliminate it?

What are all our globals?

Sorry for all the questions, but these are things which impact web devs using KMW in their own sites, and I think we may have missed some of this in the past, but want to try and manage it better in future.

Of course, if we do eliminate any globals, we should probably note that in the [migration wiki](https://github.com/keymanapp/keyman/wiki/Keyman-13.0-to-14.0-Migration), in case people used internal references in their code -- which tends to happen with open source code -- and so documenting that is just being polite when we break those non-guaranteed references.

Reply:
That was probably introduced in KMW 10.  The decision was made on this basis:

https://en.wikipedia.org/wiki/Reverse_domain_name_notation

It's basically an industry convention for namespace selection.

Since we own keyman.com, the obvious base unique namespace is `com.keyman`.  Literally all of our namespaced code is run within that joint namespace, and since we own the matching domain name, we (by convention) literally own the rights to the `com.keyman` namespace.

To eliminate it... that relates strongly to another recent discussion - the one on [converting namespaces to modules](https://docs.google.com/document/d/1WL10akQdV_-5i01nlJSyeEi8KLVGIqMldl7MMSIa3nA/edit).

So, current KMW globals:
- `KeymanWeb`
- `keyman` (We've also been setting this one since, at minimum, version 2.0.)
- `com.keyman` (Introduced during 10.0 with the start of the TypeScript conversion.)
    - As far as we're concerned, `com`'s only content is the namespace `keyman`.  Others may insert whatever they want into other `com` members with no issue; if `com` already existed, it won't be overwritten.

Reply:
In case it matters, when namespaced code like this is compiled, you get something like this:

```
var com;
(function (com) {
    var keyman;
    (function (keyman) {

       // Your code here.

    })(keyman = com.keyman || (com.keyman = {}));
})(com || (com = {}));
```

We do nothing to alter any pre-existing `com` object/namespace outside ensuring that `com.keyman` exists for our use.

Reply:
Yeah, reverse domain name notation makes a lot of sense where it is used consistently such as with Java. But it seems to me to be a fragile solution in Javascript-world. There's no reason why a website is not allowed to stomp on `window.com` any time they like and use it for their own purposes. It's not like `com` is some sort of unique identifier.

That said, I know this is an old decision. I think if/when we move to modules, we should consider eliminating that. I am comfortable with keeping both `KeymanWeb` and `keyman` as they can reasonably be expected to be unique. Of course, we'd need to assess how widely `com.keyman` is in use -- is it referenced in documentation anywhere?

Reply:
It's used fairly heavily within our code, but I don't believe that it's part of our formal documentation at this time.

Reply:
I've copied part of this conversation to the [Namespaces => Modules](https://docs.google.com/document/d/1WL10akQdV_-5i01nlJSyeEi8KLVGIqMldl7MMSIa3nA/edit#heading=h.mlyd3eklsc5x) doc you asked me to draft.  That should ensure we can revisit the notes re: `com`.